ldms/src/ldmsd/ldms_ls.c

```diff
@@ -321,18 +321,22 @@ void metric_printer(ldms_set_t s, int i)
 	enum ldms_value_type type = ldms_metric_type_get(s, i);
 	char name_str[256];
 
-	const char *metname;
+	const char *metname, *metunit;
 	if (type != LDMS_V_NONE) {
 		metname = ldms_metric_name_get(s, i);
+		metunit = ldms_metric_unit_get(s, i);
 	} else {
 		metname = "SET_ERROR";
+		metunit = "";
 	}
-	if (user_data)
-		sprintf(name_str, "%-42s 0x%" PRIx64,
```

Comment:
> @tom95858 I'm talking about (and just about) the io format to screen, not wire protocol. The currently proposed change is actual back-incompatible. Adding a -U to enable printing units only when present would preserve back-compatibility.

Hi @baallan ,

`ldms_ls` is backward-compatibility. The change passes the compatibility test on Github, which tests the result format printed by the `ldms_ls`. In more detail, `ldms_ls` in this patch printed the result the same way as the current version, including spaces between the metric name string and the metric value. You could use the printing formats on line 331 (current version) and line 333 (in the patch) as a reference.

Reply:
@nichamon 
How does replacing: 
sprintf(name_str, "%-42s 0x%" PRIx64, metname,...
with 
sprintf(name_str, "%-27s %-14s -0x%" PRIx64, metname, metunit
maintain compatibility? 
You've shortened the maximum length that will be output on a metric name (a problem we've had past problems with in things like the lustre samplers).
You've introduced a new column, which (if I parse based on whitespace (a very frequent thing) and not character-column number) will be pushing unit data into a parse routine that expects udata value or something else.
If this passed a 'back compatibility test', the test was not well formulated.

Back compatibility for _ls would be "do i get exactly the same output before and after the change".
If I want additional (new) columns, I should have to pass a new option or perhaps a multiple of an old one. -U? "-l -l"?

Reply:
@baallan, I thought your only concern is the appearance of the `ldms_ls` result on the screen when the metric unit doesn't present.

Now let me make sure that I have got all of your concerns.

- To your knowledge, there are metrics with names longer than 26 characters. (I guess this because you mentioned your past Lustre problem)
- You already have use cases that parse the `ldms_ls` results. Is this correct? Are the use cases in your test suite?

Did I miss anything?

Reply:
There are metrics with long names (>26) ; there will be metrics with longer names coming, as we expect them from C++ function names in user-defined performance sets. Truncating the names must be avoided, or the name length limit must be enforced as a failure at set definition time (adding a limit I don't recommend having).  We already implicitly (unfortunately not explicitly, which should be fixed in a separate patch) disallow using whitespace and comma in metric names to keep down-stream tools happy.

I have production scripts that parse ldms_ls output that one would expect to break under the change if samplers which naturally have units available (e.g. meminfo, lustre) displayed in the revised format automatically.  These cases are not in the ldms-static-test.sh suite. 

Most automated scripting will already have assumed the units that go with the metric they are examining via ldms, so I'm not sure I see the value in changing the default behavior of ldms_ls. Adding units as optional behavior is fine. Are the units intended to be mutable strings like udata, or to be constants like metric_name?

Reply:
Thanks for the information.

> There are metrics with long names (>26)

I will keep the space for the metric name at 42 characters then.

> Most automated scripting will already have assumed the units that go with the metric they are examining via ldms, so I'm not sure I see the value in changing the default behavior of ldms_ls.

I don't follow your logic at all. What are the automated scriptings you referred to here? How do they relate to `ldms_ls` at all?

> I have production scripts that parse ldms_ls output that one would expect to break under the change if samplers which naturally have units available (e.g. meminfo, lustre) displayed in the revised format automatically. 

There are a couple of things to address here. 

- Will `ldms_ls` report metric units when they present?
- Should we add/change the cmd-line options to configure how `ldms_ls` print metric units?
- What format will `ldms_ls` print the metric information?

My answer to the first question is yes, `ldms_ls` should report metric units if they are defined. I answered based on the `ldms_ls` program's objective, which gives set information to users, and the significance of metric units. Metric units are important information to understand a metric, which we have been missing for a long time. We design `ldms_ls` to mainly display set data on the screen.

My answer to the second issue is not to add/change options to disable/enable `ldms_ls` printing metric units. Yes, it could be a pain in the neck to update the existing scripts, primarily since this could involve changing >100 lines of code. However, this should not be a reason to complicate `ldms_ls` usage by adding/changing the cmd-line options. Also, it should not be a reason that prevents `ldms_ls` from reporting informative results, i.e., not reporting metric units.

To address the metric information format, we could decide on the format that is easy to read on the screen and avoid changing 100 lines of code in the existing tools. How large would the code change in your scripts be if the metric unit is added with a space as the delimiter?

Reply:
I came up with another format that doesn't add a new column when metric units present. The metric unit follows the metric name in the parentheses. If the metric unit doesn't exist, the result will look the same as the current version, see M2.

```
$ ldms_ls -x sock -p 10001 -l
set_0: consistent, last update: Tue Feb 16 04:51:50 2021 +0000 [1284us]
D u64        M1(Bytes)                                  3
D u64        M2                                             3

```

Reply:
This is not a satisfactory solution as we have metrics with names that already end with (xxxxxxx). please make units a separate column and make that column controllable by a format control flag.
Unix 'ls' has numerous format control flags, and there's no reason ldms_ls shouldn't.

If we're on a spree to make ldms_ls compatible and useful and extensible, please consider the example of slurm 'sacct' which supports a "machine parseable" mode. In machine parseable mode, columns are separated by a documented character (| in the case of sacct; we could use \t instead) and a corresponding column header is optionally produced using the same separator. If you aren't using "parseable" mode, you get no expectation of back compatibility.  If you add a documented parseable mode simultaneously, I withdraw any objections  about  whimsical human-oriented formatting. Notice in the sacct example that you get a "format" option to specify which columns you want (for either parsable or whimsical format).